### ISSUE TITLE
High memory consumption and fetching random proxy from list problem

### ISSUE THREAD

**Issue by enseitankado** [php-curl-class/php-curl-class#729]
When I add a large number of get requests with $multi_curl->addGet (for example 50000) a new curl instance is created for each get request. Scanning can only begin after all samples have been created. This causes excessive memory consumption. It would be a more correct approach to not create curl instances at first, but only add them to a list, and only create them during execution.

Also, the proxy list added with $multi_curl->setProxies is used randomly during scanning. This function is useless when you need to test a proxy list. Because as I said, it is randomly selected from the list. A key needs to be added to fix random proxy selection.

**Comment by zachborboa** [comment]
Hi. You can add additional requests after starting to reduce memory consumption. Take a look at this example:
https://github.com/php-curl-class/php-curl-class/blob/master/examples/multi_curl_add_curl_from_url_list.php

When using proxies, a desired proxy can be set on the individual request and a random proxy won't override the existing proxy:
https://github.com/php-curl-class/php-curl-class/blob/3ff53bfbdf83167b1b10f932c43e617f3c8005a6/src/Curl/MultiCurl.php#L1277-L1280

**Comment by enseitankado** [comment]
Hi, 

As I mentioned at the beginning; Let's say I define 100 proxies and 100 get. The library does this randomly when choosing the proxy to use for get requests. This has two downsides. First, the same proxy can be selected again, and the other can include a proxy that has never been used when 100 requests are complete. An optional key is required to use the library random proxy.

Multithread requires a certain number of get requests to start at the same time. However, the code in the example moves on to the next request after one request is complete. So the example is sequential, not multithreaded.
```php
 if (is_array($this->proxies) && $curl->getOpt(CURLOPT_PROXY) === null) { 
     $random_proxy = ArrayUtil::arrayRandom($this->proxies); 
     $curl->setProxy($random_proxy); 
 } 
```

**Comment by zachborboa** [comment]
The issue with high memory use should be resolved by using a pattern like this:
https://github.com/php-curl-class/php-curl-class/blob/master/examples/multi_curl_add_curl_from_url_list.php

To use a unique random proxy, something like this could work:
https://github.com/php-curl-class/php-curl-class/blob/master/examples/multi_curl_get_with_new_random_unique_proxy.php

**Comment by enseitankado** [comment]
The idea of destroying the once used proxy inside the beforeSend procedure is nice.
Unfortunately I did this by modifying the library itself, which would create additional work in updating the library. Now I'm familiar with the beforeSend function. Thanks for the illustrative sample program.

Funny thing is, I realized that I persisted in my question about memory consumption. When I looked a little more carefully, I decided that the example you gave worked correctly.

I see you are a few steps ahead of me. Thank you for your patience.
